openedx/features/pakx/lms/overrides/utils.py

```diff
@@ -392,6 +397,88 @@ def get_course_progress_percentage(request, course_key):
     return format((total_completed_blocks / total_blocks) * 100, '.0f') if total_blocks > 0 else total_blocks
 
 
+def check_and_unlock_user_milestone(user, course_key):
+    """Checks if pre req for locked subsection have been completed."""
+
+    course_key = CourseKey.from_string(course_key)
+    milestones = get_prerequisites(course_key)
+    final_milestone, prereq_for_final = _get_milestones(milestones)
+
+    if not final_milestone:
+        log.info('No final milestone found for course:{}'.format(course_key))
+        return
+
+    try:
+        if milestones_api.user_has_milestone(model_to_dict(user), prereq_for_final):
+            _unlock_or_add_unlock_date(user, course_key, final_milestone)
+    except InvalidMilestoneException:
+        log.info('User:{} for course:{} have milestone:{}'.format(user.email, course_key, prereq_for_final))
+
+
+def _unlock_or_add_unlock_date(user, course_key, final_milestone):
+    """Check and unlock subsection if unlock date has been specified and fulfilled."""
+
+    from .models import CourseProgressStats
+
+    try:
+        date_to_unlock, course_progress = _get_course_progress(user, course_key, CourseProgressStats)
+    except (CourseOverviewContent.DoesNotExist, CourseProgressStats.DoesNotExist):
+        log.info(
+            'Course Progress stats or Course Overview does not exist for user:{} and course:{}'.format(
+                user.email,
+                course_key
+            )
+        )
+        return
+
+    if not course_progress.unlock_subsection_on:
+        course_progress.unlock_subsection_on = date_to_unlock
+        course_progress.save()
```

Comment:
use update_fields while calling save. it only updates the desired columns instead of updating the whole object